### PR TITLE
Restore early lake labels, add kind_detail for lakes and etc

### PR DESCRIPTION
**Pull request by nvkelso**
Connects with https://github.com/tilezen/vector-datasource/issues/2003 and followup of https://github.com/tilezen/vector-datasource/pull/2010 to restore "lake" water labels at low zooms.

Also non-breaking change implementation of #984 for the v1.x series (which should stay open as a separate breaking change for v2.x series).

- [x] Update tests
- [x] Update docs

### CODE REVIEW

**Comment by peitili** [comment]
there is a linter violations, you might want to fix them using the pre-commit

**Comment by peitili** [comment]
I am not sure why but the integration tests always failed with context deadline exceed.

File: integration-test/628-standardize-water-kinds.py [hunk 1]
```diff
@@ -23,3 +23,33 @@ def test_reservoir(self):
         self.assert_has_feature(
             16, 10481, 25324, 'water',
             {'kind': 'lake', 'reservoir': True})
+
+    # this test is for https://github.com/tilezen/vector-datasource/pull/2047
+    def test_water_reservoir(self):
```
Comment (peitili) [inline integration-test/628-standardize-water-kinds.py:28]: @nvkelso please advice if we need some non-fake data and appreciate it if you can let me know which ones.

Also please advice whether we should add these tests here in this file or other issue number.

**Review by nvkelso — COMMENTED** [review]
Please also add a test for the low zoom removals.

Clear Lake in California: [-4046481](https://www.openstreetmap.org/relation/4046481) with an area of `265548000`.

That features name was being stripped until zoom 12 in error, but with this change it should be visible in a zoom from zoom 5+ depending on it's area, which looks like results in a min_zoom of 6. Since it's OSM it'll not actually be in zoom 6, so test a zoom 8 tile and see if that feature is there with that min_zoom. 

This is the 256px tile I'd expect to see such a label in:

- https://tile.openstreetmap.org/10/162/391.png

File: integration-test/628-standardize-water-kinds.py [hunk 2]
```diff
@@ -23,3 +23,33 @@ def test_reservoir(self):
         self.assert_has_feature(
             16, 10481, 25324, 'water',
             {'kind': 'lake', 'reservoir': True})
+
+    # This test is for https://github.com/tilezen/vector-datasource/pull/2047
+    def test_water_reservoir(self):
+        # Francisco Reservoir, SF
+        self.generate_fixtures(dsl.way(386662458, wkt_loads('POLYGON ((-122.421611012467 37.80395991667648, -122.420227786592 37.80413210674329, -122.420159873957 37.80408277780808, -122.420117832801 37.80383748120349, -122.42154480663 37.80365748294528, -122.421611012467 37.80395991667648))'), {
```
Comment (nvkelso) [inline integration-test/628-standardize-water-kinds.py:30]: Is this the right OSM ID?

File: integration-test/628-standardize-water-kinds.py [hunk 3]
```diff
@@ -23,3 +23,33 @@ def test_reservoir(self):
...
+                               u'way_area': u'6936.45', u'source': u'openstreetmap.org', u'water': u'reservoir', u'name': u'Francisco Reservoir', u'covered': u'no'}))
+
+        self.assert_has_feature(
+            16, 10481, 25324, 'water',
+            {'kind': 'lake', 'reservoir': True, 'kind_detail': 'lake'})
+
+    # This test is for https://github.com/tilezen/vector-datasource/pull/2047
+    def test_water_lagoon(self):
+        # Francisco Reservoir, SF
+        self.generate_fixtures(dsl.way(386662458, wkt_loads('POLYGON ((-122.421611012467 37.80395991667648, -122.420227786592 37.80413210674329, -122.420159873957 37.80408277780808, -122.420117832801 37.80383748120349, -122.42154480663 37.80365748294528, -122.421611012467 37.80395991667648))'), {
```
Comment (nvkelso) [inline integration-test/628-standardize-water-kinds.py:40]: Seems like wrong OSM ID.

File: integration-test/628-standardize-water-kinds.py [hunk 4]
```diff
@@ -23,3 +23,33 @@ def test_reservoir(self):
...
+                               u'way_area': u'6936.45', u'source': u'openstreetmap.org', u'water': u'lagoon', u'name': u'Francisco Reservoir', u'covered': u'no'}))
+
+        self.assert_has_feature(
+            16, 10481, 25324, 'water',
+            {'kind': 'lake', 'alkaline': True, 'kind_detail': 'lake'})
+
+    # This test is for https://github.com/tilezen/vector-datasource/pull/2047
+    def test_water_ditch(self):
+        # Francisco Reservoir, SF
+        self.generate_fixtures(dsl.way(386662458, wkt_loads('POLYGON ((-122.421611012467 37.80395991667648, -122.420227786592 37.80413210674329, -122.420159873957 37.80408277780808, -122.420117832801 37.80383748120349, -122.42154480663 37.80365748294528, -122.421611012467 37.80395991667648))'), {
```
Comment (nvkelso) [inline integration-test/628-standardize-water-kinds.py:50]: Wrong OSM ID. If it's not relevant just make it be `1` instead.

File: integration-test/628-standardize-water-kinds.py [hunk 5]
```diff
@@ -23,3 +23,33 @@ def test_reservoir(self):
...
+                               u'way_area': u'6936.45', u'source': u'openstreetmap.org', u'water': u'ditch', u'name': u'Francisco Reservoir', u'covered': u'no'}))
```
Comment (nvkelso) [inline integration-test/628-standardize-water-kinds.py:51]: I would expect another tag here: 

`'natural': u'water'`

File: integration-test/628-standardize-water-kinds.py [hunk 6]
```diff
@@ -23,3 +23,33 @@ def test_reservoir(self):
...
+
+        self.assert_has_feature(
+            16, 10481, 25324, 'water',
+            {'kind': 'lake', 'kind_detail': 'ditch'})
```
Comment (nvkelso) [inline integration-test/628-standardize-water-kinds.py:55]: With the new `natural=water` tag addition above, I'd expect to see output more like:

>            {'kind': 'water', 'kind_detail': 'ditch'})

File: integration-test/628-standardize-water-kinds.py [hunk 7]
```diff
@@ -23,3 +23,33 @@ def test_reservoir(self):
...
+
+    # This test is for https://github.com/tilezen/vector-datasource/pull/2047
+    def test_water_lagoon(self):
+        # Francisco Reservoir, SF
+        self.generate_fixtures(dsl.way(386662458, wkt_loads('POLYGON ((-122.421611012467 37.80395991667648, -122.420227786592 37.80413210674329, -122.420159873957 37.80408277780808, -122.420117832801 37.80383748120349, -122.42154480663 37.80365748294528, -122.421611012467 37.80395991667648))'), {
+                               u'way_area': u'6936.45', u'source': u'openstreetmap.org', u'water': u'lagoon', u'name': u'Francisco Reservoir', u'covered': u'no'}))
+
+        self.assert_has_feature(
+            16, 10481, 25324, 'water',
+            {'kind': 'lake', 'alkaline': True, 'kind_detail': 'lake'})
```
Comment (nvkelso) [inline integration-test/628-standardize-water-kinds.py:45]: With the new `natural=water` tag addition above, I'd expect to see output more like:

>             {'kind': 'water', 'alkaline': True, 'kind_detail': 'lake'})

File: integration-test/628-standardize-water-kinds.py [hunk 8]
```diff
@@ -23,3 +23,33 @@ def test_reservoir(self):
...
+
+        self.assert_has_feature(
+            16, 10481, 25324, 'water',
+            {'kind': 'lake', 'reservoir': True, 'kind_detail': 'lake'})
+
+    # This test is for https://github.com/tilezen/vector-datasource/pull/2047
+    def test_water_lagoon(self):
+        # Francisco Reservoir, SF
+        self.generate_fixtures(dsl.way(386662458, wkt_loads('POLYGON ((-122.421611012467 37.80395991667648, -122.420227786592 37.80413210674329, -122.420159873957 37.80408277780808, -122.420117832801 37.80383748120349, -122.42154480663 37.80365748294528, -122.421611012467 37.80395991667648))'), {
+                               u'way_area': u'6936.45', u'source': u'openstreetmap.org', u'water': u'lagoon', u'name': u'Francisco Reservoir', u'covered': u'no'}))
```
Comment (nvkelso) [inline integration-test/628-standardize-water-kinds.py:41]: I would expect another tag here: 

`'natural': u'water'`

File: integration-test/628-standardize-water-kinds.py [hunk 9]
```diff
@@ -23,3 +23,33 @@ def test_reservoir(self):
         self.assert_has_feature(
             16, 10481, 25324, 'water',
             {'kind': 'lake', 'reservoir': True})
+
+    # This test is for https://github.com/tilezen/vector-datasource/pull/2047
+    def test_water_reservoir(self):
+        # Francisco Reservoir, SF
+        self.generate_fixtures(dsl.way(386662458, wkt_loads('POLYGON ((-122.421611012467 37.80395991667648, -122.420227786592 37.80413210674329, -122.420159873957 37.80408277780808, -122.420117832801 37.80383748120349, -122.42154480663 37.80365748294528, -122.421611012467 37.80395991667648))'), {
+                               u'way_area': u'6936.45', u'source': u'openstreetmap.org', u'water': u'reservoir', u'name': u'Francisco Reservoir', u'covered': u'no'}))
```
Comment (nvkelso) [inline integration-test/628-standardize-water-kinds.py:31]: I would expect another tag here: 

`'natural': u'water'`

File: integration-test/628-standardize-water-kinds.py [hunk 10]
```diff
@@ -23,3 +23,33 @@ def test_reservoir(self):
...
+
+        self.assert_has_feature(
+            16, 10481, 25324, 'water',
+            {'kind': 'lake', 'reservoir': True, 'kind_detail': 'lake'})
```
Comment (nvkelso) [inline integration-test/628-standardize-water-kinds.py:35]: With the new natural=water tag addition above, I'd expect to see output more like:

>             {'kind': 'water', 'reservoir': True, 'kind_detail': 'lake'})

File: integration-test/628-standardize-water-kinds.py (hunk 1 above)
Comment (nvkelso) [reply integration-test/628-standardize-water-kinds.py:28]: It's ok, but a separate new test would be better since you're not otherwise modifying any of the other tests in this old file.

Comment (nvkelso) [reply integration-test/628-standardize-water-kinds.py:28]: I mentioned in the general PR, but responding directly here too. 

Please also add a test for the low zoom removals.

Clear Lake in California: [-4046481](https://www.openstreetmap.org/relation/4046481) with an area of 265548000. You can copy the props but just use DSL to make a box that's as large as that area instead of copying all the detailed polygon in.

That features name was being stripped until zoom 12 in error, but with this change it should be visible in a zoom from zoom 5+ depending on it's area, which looks like results in a min_zoom of 6. Since it's OSM it'll not actually be in zoom 6, so test a zoom 10 tile and see if that feature is there with that min_zoom.

This is the 256px tile I'd expect to see such a label in:

https://tile.openstreetmap.org/10/162/391.png

File: integration-test/628-standardize-water-kinds.py (hunk 2 above)
Comment (nvkelso) [reply integration-test/628-standardize-water-kinds.py:30]: If it's made up, then use `1` instead.

File: integration-test/628-standardize-water-kinds.py (hunk 3 above)
Comment (nvkelso) [reply integration-test/628-standardize-water-kinds.py:40]: If it's made up, then use `1` instead.

File: integration-test/628-standardize-water-kinds.py (hunk 1 above)
Comment (nvkelso) [reply integration-test/628-standardize-water-kinds.py:28]: Please also add a test for the low zoom removals.

Clear Lake in California: [-4046481](https://www.openstreetmap.org/relation/4046481) with an area of 265548000. You can copy the props but just use DSL to make a box that's as large as that area instead of copying all the detailed polygon in.

That features name was being stripped until zoom 12 in error, but with this change it should be visible in a zoom from zoom 5+ depending on it's area, which looks like results in a min_zoom of 6. Since it's OSM it'll not actually be in zoom 6, so test a zoom 10 tile and see if that feature is there with that min_zoom.

This is the 256px tile I'd expect to see such a label in:

https://tile.openstreetmap.org/10/162/391.png

Comment (nvkelso) [reply integration-test/628-standardize-water-kinds.py:28]: (It's OK but not great to edit the old test – but since you're not fixing anything in the old test, it's better to make a new file with the PR number in the name.)

File: integration-test/2047-restore-early-lake-labels.py [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+            u'covered': u'no'}))
+
+        self.assert_has_feature(
+            16, 10481, 25324, 'water',
+            {
+                'kind': 'water',
+                'kind_detail': 'ditch'
+            })
+
+    def test_name_not_drop(self):
```
Comment (peitili) [inline integration-test/2047-restore-early-lake-labels.py:93]: Traceback (most recent call last):
  File "/Users/pli2/Snapchat/Dev/vector-datasource/integration-test/2047-restore-early-lake-labels.py", line 116, in test_name_drop
    'min_zoom': 6.0,
  File "/Users/pli2/Snapchat/Dev/vector-datasource/integration-test/__init__.py", line 1492, in assert_has_feature
    self.test_instance.assert_has_feature(z, x, y, layer, props)
  File "integration-test/__init__.py", line 1325, in assert_has_feature
    self.assertions.assert_has_feature(z, x, y, layer, props)
  File "integration-test/__init__.py", line 1154, in assert_has_feature
    'layer %r was empty)' % (properties, layer))
AssertionError: Did not find feature including properties {'kind_detail': 'lake', 'kind': 'water', 'min_zoom': 6.0, 'name': 'Clear Lake'} (because layer 'water' was empty)

File: integration-test/2047-restore-early-lake-labels.py [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+        import dsl
+
+        for z in range(6, 13):
```
Comment (peitili) [inline integration-test/2047-restore-early-lake-labels.py:96]: @nvkelso I found that this test is strange it can only pass if the zoom is in this range

if I specify zoom <=5 it will complain the following, hmm why? I thought the water layer should still exist and only the name label shouldn't be there.

```
Traceback (most recent call last):
  File "/Users/pli2/Snapchat/Dev/vector-datasource/integration-test/2047-restore-early-lake-labels.py", line 116, in test_name_drop
    'min_zoom': 6.0,
  File "/Users/pli2/Snapchat/Dev/vector-datasource/integration-test/__init__.py", line 1492, in assert_has_feature
    self.test_instance.assert_has_feature(z, x, y, layer, props)
  File "integration-test/__init__.py", line 1325, in assert_has_feature
    self.assertions.assert_has_feature(z, x, y, layer, props)
  File "integration-test/__init__.py", line 1154, in assert_has_feature
    'layer %r was empty)' % (properties, layer))
AssertionError: Did not find feature including properties {'kind_detail': 'lake', 'kind': 'water', 'min_zoom': 6.0, 'name': 'Clear Lake'} (because layer 'water' was empty)
```

for any zoom >= 13 it will throw this error

```
FAIL: test_name_drop (integration-test.2047-restore-early-lake-labels.WaterKinds)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/pli2/Snapchat/Dev/vector-datasource/integration-test/2047-restore-early-lake-labels.py", line 100, in test_name_drop
    dsl.way(1, dsl.box_area(z, x, y, area, include_boundary=True), {
  File "/Users/pli2/Snapchat/Dev/vector-datasource/integration-test/dsl.py", line 184, in box_area
    _check(centre_x - 0.5 * size_x),
  File "/Users/pli2/Snapchat/Dev/vector-datasource/integration-test/dsl.py", line 180, in _check
    assert -half_earth_circum <= coord <= half_earth_circum
AssertionError
```

File: integration-test/2047-restore-early-lake-labels.py [hunk 13]
```diff
@@ -0,0 +1,117 @@
+# -*- encoding: utf-8 -*-
+import dsl
+from shapely.wkt import loads as wkt_loads
+
+from . import FixtureTest
+
+
+class BoundsTest(FixtureTest):
```
Comment (peitili) [inline integration-test/2047-restore-early-lake-labels.py:8]: this is unnecessary, should remove.

File: integration-test/628-standardize-water-kinds.py (hunk 1 above)
Comment (peitili) [reply integration-test/628-standardize-water-kinds.py]: fixed in the latest commit

File: integration-test/2047-restore-early-lake-labels.py [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+
+        for z in range(6, 13):
+            x, y = (3, 4)
+            area = 265548000
+
+            self.generate_fixtures(
+                dsl.way(1, dsl.box_area(z, x, y, area, include_boundary=True), {
+                    'name': u'Clear Lake',
+                    'natural': u'water',
+                    'type': u'multipolygon',
```
Comment (nvkelso) [inline integration-test/2047-restore-early-lake-labels.py:104]: Please remove this, just confuses things since we're faking a box for the shape.

File: integration-test/2047-restore-early-lake-labels.py [hunk 15]
```diff
@@ -0,0 +1,117 @@
...
+                    'source': u'openstreetmap.org',
+                }),
+            )
+
+            self.assert_has_feature(
+                z, x, y, 'water', {
+                    'kind': 'water',
+                    'kind_detail': 'lake',
+                    'name': 'Clear Lake',
+                    'min_zoom': 6.0,
```
Comment (nvkelso) [inline integration-test/2047-restore-early-lake-labels.py:116]: For the generated label point geometry (the cruz of this PR), you want to also test for:

>                     'label_placement': True,

Otherwise you're testing the polygon which has it's own min_zoom range.

To clarify... the polygon should be visible from zoom 6 conceptually based on it's area grading of `265548000` sq meters in https://github.com/tilezen/vector-datasource/blob/master/yaml/water.yaml#L23-L46.

```
        - [ 5,   500000000 ]
        - [ 6,   200000000 ]
        - [ 7,    50000000 ]
```

For the label we do a remapping of the min_zoom based on area grading of `265548000` sq metershere: https://github.com/tilezen/vector-datasource/blob/master/queries.yaml#L843-L878, so I expect min_zoom 9 for the label.

```
           (8,     500000000),
            (9,     200000000),
            (10,     40000000),
```

Since we don't transition from NE to OSM varies based on layer, but zoom 4 presumably for water from the osm polygons table:

https://github.com/tilezen/vector-datasource/blob/master/queries.yaml#L23-L26

So to recap: 
- the lake polygon shouldn't be visible until zoom 6, and it won't have a name property ever
- the lake label shouldn't be visible until zoom 9 (but wasn't visible until zoom 12 in error before), so we are testing for it's presence at zooms 9, 10, and 11 (and that it's not present at 8, and still present at zoom 12).

Comment (peitili) [reply integration-test/2047-restore-early-lake-labels.py]: @nvkelso please see the latest commit about that integration test, the behavior is not what you described here. 

what the integ test told us are:

1. We are seeing `name` property appear since zoom 6 for the lake polygon
2. We are seeing label_placement since zoom 6

File: integration-test/2047-restore-early-lake-labels.py [hunk 16]
```diff
@@ -0,0 +1,123 @@
...
+                self.assert_has_feature(
+                    z, x, y, 'water', {
+                        'kind': 'water',
+                        'kind_detail': 'lake',
+                        'name': 'Clear Lake',
+                        'min_zoom': 6.0,
+                    })
+
+            # the following assert will start to fail on z == 6
+            if z < 6:
```
Comment (peitili) [inline integration-test/2047-restore-early-lake-labels.py:108]: @nvkelso  starting z==6 there are feature in the water layer


`{'label_placement': True, 'kind': 'water', 'collision_rank': 487, 'name': 'Clear Lake', 'area': 265548000, 'sort_rank': 204, 'kind_detail': 'lake', 'source': 'openstreetmap.org', 'min_zoom': 6.0, 'wikidata_id': 'Q1099503', 'id': 1}`

File: integration-test/2047-restore-early-lake-labels.py (hunk 13 above)
Comment (nvkelso) [reply integration-test/2047-restore-early-lake-labels.py]: Go for it!

File: integration-test/2047-restore-early-lake-labels.py [hunk 17]
```diff
@@ -0,0 +1,123 @@
...
+                                                {
+                                                    'kind': 'water',
+                                                    'kind_detail': 'lake',
+                                                })
+            if z >= 6:
+                self.assert_has_feature(
+                    z, x, y, 'water', {
+                        'kind': 'water',
+                        'kind_detail': 'lake',
+                        'name': 'Clear Lake',
```
Comment (nvkelso) [inline integration-test/2047-restore-early-lake-labels.py:103]: The name shouldn't be present, per https://github.com/tilezen/vector-datasource/blob/master/queries.yaml#L898-L911.

Why isn't this test failing?

File: integration-test/2047-restore-early-lake-labels.py (hunk 16 above)
Comment (nvkelso) [reply integration-test/2047-restore-early-lake-labels.py:108]: Hmm. For the label we do a remapping of the min_zoom based on area grading of 265548000 sq metershere: https://github.com/tilezen/vector-datasource/blob/master/queries.yaml#L843-L878, so I expect min_zoom 9 for the label.

```
           (8,     500000000),
            (9,     200000000),
            (10,     40000000),
```

File: integration-test/2047-restore-early-lake-labels.py [hunk 18]
```diff
@@ -0,0 +1,123 @@
...
+                                                })
+
+            # if z >= 9:
+            #     self.assert_has_feature(
+            #         z, x, y, 'water', {
+            #             'kind': 'water',
+            #             'kind_detail': 'lake',
+            #             'name': 'Clear Lake',
+            #             'label_placement': True,
+            #             'min_zoom': 6.0,
```
Comment (nvkelso) [inline integration-test/2047-restore-early-lake-labels.py:122]: The label's min_zoom should be `9.0` not `6.0`.

File: integration-test/2047-restore-early-lake-labels.py (hunk 16 above)
Comment (nvkelso) [reply integration-test/2047-restore-early-lake-labels.py:108]: This looks like an existing bug in that logic, as the last build shows the label for Clear Lake is in fact 6.0 not 9.0 like intended. Let's fix that.

Comment (nvkelso) [reply integration-test/2047-restore-early-lake-labels.py:108]: The area threshold numbers look fine to me. Is this where clause is malformed?

> (properties is not None and properties.get('label_placement') is True)

File: integration-test/2047-restore-early-lake-labels.py (hunk 15 above)
Comment (peitili) [reply integration-test/2047-restore-early-lake-labels.py]: there is a bug in the yaml file processing in the update_min_zoom function, just fixed that and then fixed the integration tests

File: vectordatasource/transform.py [hunk 19]
```diff
@@ -9471,6 +9470,9 @@ def update_min_zoom(ctx):
         local = defaultdict(lambda: None)
         local.update(props)
         local['zoom'] = zoom
+        # this is to make the name `properties` visible in the queries.yaml's
+        # where clause
+        local['properties'] = props
```
Comment (nvkelso) [inline vectordatasource/transform.py:9475]: This may impact build time, but generally makes things more consistent and obvious.

**Review by nvkelso — COMMENTED** [review]
Final set of code, tests, and docs LGTM, including Q&A of a mini build.

@peitili this is ready to merge

**Review by peitili — APPROVED** [review]
